### PR TITLE
Deprecate enqueue!, dequeue! and peek for PriorityQueue

### LINKED ISSUE #296: Use `push!` and `pop!` for `PriorityQueue` etc
We should use `push!` and `pop!` instead of `enqueue!` and `dequeue!`, 
so that different data structures are interchangeable without changing code.

cf. #289

### PR DESCRIPTION
**Pull request by jonas-schulze**
I took a stab at the remaining part of #296. Please note my [comment on discourse](https://discourse.julialang.org/t/taking-push-and-pop-seriously/34326/20?u=jonas-schulze) on instead using `put!` and `take!`, but I think a consistent API is more important now.

If the refactor of `popfirst!` to match `eltype` went too far (*breaking change*), I can drop the corresponding commit. Also, I for now deprecated `enqueue(pq, key, val)` in favor of `Base.push!(pq, key=>val)` instead of `Base.insert!(pq, key, val)`. Less is more, I think. `Dict` doesn't have `push!` nor `insert!`.

This is ready for comments, but there are some parts missing:

- [x] Decide on refactor of `popfirst!` (Edit: it's fine)
- [x] Decide on `push!` vs `insert!` (Edit: it's ok as it is)
- [x] Update docs

Close #296

### CODE REVIEW

**Comment by jonas-schulze** [comment]
The reported coverage loss is in `src/splay_tree.jl`, which I didn't even touch. 👀 
https://app.codecov.io/gh/JuliaCollections/DataStructures.jl/compare/771/changes

**Comment by jonas-schulze** [comment]
@oxinabox bump 🙂

File: src/priorityqueue.jl
Comment (oxinabox) [inline src/priorityqueue.jl:283]: What do you think of varying the normal rule against using import?
```suggestion
if isdefined(Base, :popat!)  # We will overload if it is defined, else we define on our own
    import popat!
end

function popat!(pq::PriorityQueue, key)
```
I think it might be the lesser of two evils here?

File: src/DataStructures.jl
Comment (oxinabox) [inline src/DataStructures.jl:10]: What do you think of:
```suggestion
    export popat!  # Export for old version of julia where Base doesn't export this
```

I think it should work since it is either exporting something new (fine)
or it is a no-op by exporting something already exported

File: test/test_priority_queue.jl
Comment (oxinabox) [inline test/test_priority_queue.jl:114]: Probably out of scope for this PR, but this use of `peek` should be replaces with `first`.
Which will
A) Nicely match to `popfirst`
B) Match `first` on a `Dict`
C) be consistent with `Base` using `peek` only for IO streams

**Review by oxinabox — COMMENTED** [review]
Nice work.
This looks good to me.
Just a few comments that I will wait for you to reply to before I merge

**Review by oxinabox — APPROVED** [review]

File: src/priorityqueue.jl
Comment (jonas-schulze) [reply src/priorityqueue.jl:283]: I like what you proposed in https://github.com/JuliaCollections/DataStructures.jl/pull/771#discussion_r751502217, but I don't have a strong preference here. Which two evils do you mean here?

Comment (oxinabox) [reply src/priorityqueue.jl]: Evil 1: is `import Foo: bar` because it allowed unqualified overloads
Evil 2: is this code is kinda complex for what it it does.

I think yes let's conditionally import it

**Review by oxinabox — APPROVED** [review]

**Comment by oxinabox** [comment]
thanks!

**Pull request merged**
